styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -251,17 +248,18 @@ private void updateResourceCounters(StorageTransaction tx, Event event,
       tryUpdatingCounter(currentRunState, tx, nextRunState.data().resourceIds().get());
     }
 
+    // decrement counters if transitioning from a state that consumes resources
+    // to a state that doesn't consume any resources
     if (isConsumingResources(currentRunState.state())
-        && !isConsumingResources(nextRunState.state())
-        && nextRunState.data().resourceIds().isPresent()) {
-      // decrement counters if transitioning from a state that consumes resources
-      // to a state that doesn't consume any resources
-      for (String resource : nextRunState.data().resourceIds().get()) {
-        tx.updateCounter(shardedCounter, resource, -1);
+        && !isConsumingResources(nextRunState.state())) {
+      if (nextRunState.data().resourceIds().isPresent()) {
+        for (String resource : nextRunState.data().resourceIds().get()) {
+          tx.updateCounter(shardedCounter, resource, -1);
+        }
+      } else {
+        LOG.error("Resource ids are missing for {} when transitioning from {} to {}.",
```

Comment:
I think it is acceptable to have RunStates with no resource id. Even if practically we should always have at least global resource and some decorators, those are not a mandatory configuration in OSS.

Reply:
The logic of persisting the set is meant to be consistent, even in OSS. We aren't checking if it has elements in it, just checking the optional that it `isPresent`.